django/contrib/auth/hashers.py

```diff
@@ -512,6 +513,84 @@ class BCryptPasswordHasher(BCryptSHA256PasswordHasher):
     digest = None
 
 
+class ScryptPasswordHasher(BasePasswordHasher):
+    """
+    Secure password hashing using the Scrypt algorithm
+
+    The result is a 64 byte binary string.
+    """
+    algorithm = "scrypt"
+    work_factor = 16384
+    block_size = 8
+    parallelism = 1
+    maxmem = 0
+
+    def encode(self, password, salt, n=work_factor, r=block_size, maxmem=maxmem):
+        assert password is not None
+        assert salt and '$' not in salt
+        hash = scrypt(
+            password, salt, n=n, r=r, p=self.parallelism, maxmem=maxmem
+        )
+        hash = base64.b64encode(hash).decode('ascii').strip()
+        return "%s$%d$%s$%s$%d$%d$%d" % (
+            self.algorithm, n, salt, hash, r,
+            self.parallelism, maxmem
+        )
+
+    def decode(self, encoded):
+        algorithm, work_factor, salt, hash, block_size, parallelism, maxmem = encoded.split('$', 6)
+        assert algorithm == self.algorithm
+        return {
+            'algorithm': algorithm,
+            'hash': hash,
+            'work_factor': int(work_factor),
+            'salt': salt,
+            'block_size': int(block_size),
+            'parallelism': int(parallelism),
+            'maxmem': int(maxmem)
+        }
+
+    def verify(self, password, encoded):
+        decoded = self.decode(encoded)
+        encoded_2 = self.encode(
+            password, decoded['salt'], decoded['work_factor'],
+            decoded['block_size']
+        )
+        return constant_time_compare(encoded, encoded_2)
+
+    def safe_summary(self, encoded):
+        decoded = self.decode(encoded)
+        return {
+            _('algorithm'): decoded['algorithm'],
+            _('work factor'): decoded['work_factor'],
+            _('block size'): decoded['block_size'],
+            _('parallelism'): decoded['parallelism'],
+            _('maxmem'): decoded['maxmem'],
+            _('salt'): mask_hash(decoded['salt']),
+            _('hash'): mask_hash(decoded['hash'])
+        }
+
+    def must_update(self, encoded):
+        decoded = self.decode(encoded)
+        return (
+            decoded['work_factor'] != self.work_factor or
+            decoded['block_size'] != self.block_size or
+            decoded['parallelism'] != self.parallelism or
+            decoded['maxmem'] != self.maxmem
+        )
+
+    def harden_runtime(self, password, encoded):
```

Comment:
My main worry here is: Is this correct and does it make sense to implement for such a complex hasher (notably we already have others where we argue it is simply not possible in a sensible way).

Since scrypt can raise errors like this:
> ValueError: Invalid parameter combination for n, r, p, maxmem.

I am wondering if `must_update` couldn't also trigger this condition. Or can we always calculate `extra_iterations` and `extra_block` and be sure that the combinations are valid?

Reply:
> My main worry here is: Is this correct and does it make sense to implement for such a complex hasher (notably we already have others where we argue it is simply not possible in a sensible way).
> 
> Since scrypt can raise errors like this:
> 
> > ValueError: Invalid parameter combination for n, r, p, maxmem.
> 
> I am wondering if `must_update` couldn't also trigger this condition. Or can we always calculate `extra_iterations` and `extra_block` and be sure that the combinations are valid?

You're right parameters may no be valid, e.g.
```
self.work_factor = 2 **14
decoded['work_factor'] = 2 ** 11
```
both are a power of 2, however `extra_iterations = 14336` is not and raises `ValueError: n must be a power of 2`.